src/containers/ListingPage/ListingPageCarousel.js

```diff
@@ -496,17 +499,25 @@ const EnhancedListingPage = props => {
     );
   }
 
-  const currentUser = props.currentUser;
   const isPrivateMarketplace = config.accessControl.marketplace.private === true;
   const isUnauthorizedUser = currentUser && !isUserAuthorized(currentUser);
+  const hasNoViewingRights = currentUser && !hasReadPermission(currentUser);
   const hasUserPendingApprovalError = isErrorUserPendingApproval(showListingError);
+
   if ((isPrivateMarketplace && isUnauthorizedUser) || hasUserPendingApprovalError) {
     return (
       <NamedRedirect
         name="NoAccessPage"
         params={{ missingAccessRight: NO_ACCESS_PAGE_USER_PENDING_APPROVAL }}
       />
     );
+  } else if (hasNoViewingRights && isForbiddenError(showListingError)) {
```

Comment:
My assumption is that this should be || not &&

Reply:
You could also add a check for "isPrivateMarketplace &&" as above.

The point is that we want to make sure that on what conditions these redirections happen. Devs (future us & customizers) should not prevent SSR on listing pages on public marketplaces.

Reply:
If this is hasNoViewingRights || isForbiddenError, then it shows the NoAccessPage also when users are trying to reach their own listings.

By default, the .duck.js file is defined to call ownListings endpoint for all listings when a user has restricted viewing rights, and the endpoint throws 403 for everyone else's listings but gives 200 for the current user's own listings. So the combination of noViewingRights && 403 means that the user has tried to view someone else's listings, but they are getting 403 from the ownListings endpoint.

If there's some other way to handle allowing own listings but not other folks' listings, and without fetching the listing first, I can also use that in the .duck.js file.

Reply:
There is error code: permission-denied-read

Reply:
If we use the ownListings.show endpoint in ListingPage.duck.js file, the 403 does not come back as permission-denied-read, because the 403 is caused by a non-author attempting ownListings.show, and the error message is `forbidden`. If we use listings.show endpoint in ListingPage.duck.js, it gives 403 permission-denied-read on all listings, including the user's own, so the isOwnListing check cannot be applied after the SDK call.

However, since read rights only apply to private marketplaces, I could set the ListingPage.duck.js hasViewingRights condition to 

`const hasViewingRights = !isPrivateMarketplace || (isPrivateMarketplace && currentUser && hasPermissionToViewData(currentUser));`

and ListingPage[variant].js hasNoViewingRights to

`const hasNoViewingRights = isPrivateMarketplace && currentUser && !hasPermissionToViewData(currentUser);`

This way, the ownListings.show endpoint (in duck.js) and hasNoViewingRights would explicitly only get applied on a private marketplace. That's what happens in practice anyway, as Console doesn't allow revoking read rights on public marketplaces, but it's also clear for customizers.

Would something like this work?

This would address the issues you listed:

- public marketplace: SSR must work without currentUser entity => hasViewingRights check does not apply
  - The aforementioned error: redirection => this won't ever happen
  - Happy case: by-pass this redirection => this always happens
- private marketplace: CSR only => hasViewingRights check applies in duck.js, currentUser already exists
  - own listings: by-pass this redirection => always shown, whether .duck.js file uses listings.show or ownListings.show
  - someone elses listings => viewing rights checked in duck.js
      - has viewing rights: by-pass this redirection => uses listings.show endpoint, no error
      - no viewing rights: redirection => uses ownListings.show endpoint, gets 403

Reply:
ListingPage & ProfilePage are part of SSR and because the knowledge of private vs public comes from an asset (that can be stale because assets are cached for 5 min in production), we need to consider how SSR behaves on profile, listing & search pages - even if the marketplace is turned to private mode.

If I read the code correctly, the SSR during those stale moments will fetch user on profile page and listing on listing page. I doubt we can avoid it.

> Side note: I think I made a mistake earlier, when I added fetchCurrentUser to the loadData of this page. Luckily, full page loading doesn't fetch currentUser because isAuthenticated flag is false at that point of data loading and the fetch is not done on the server.

Reply:
Listing pages NoAccessPage condition tested with SSR-setup (i.e. private marketplace set to false on asset and no current user) and set to

`} else if (hasNoViewingRights && isForbiddenError(showListingError)) {`

=> this condition shows NoAccessPage for missing viewing rights – if hasViewingRights resolves to false in .duck.js, the endpoint that gets used is ownListings, and the error from that endpoint is `forbidden` for no current user. Removed the privateMarketplace check so this catches SSR-situations as well.